From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 15 Apr 2026 01:12:53 -0400
Subject: [PATCH] Improve logger/terminal shutdown

Note: improper solution, needs to research later.

Move logger and terminal shutdown processes to the correct stage,
which is at the end of the server exit and before the System#exit.

The original way relies on the finalization.
And the finalization is untrustworthy, see JEP 421: https://openjdk.org/jeps/421
The System#exit may also interrupt the normal closure.

The race still exists, and the logging still happens after the terminal is shut down.
This problem is entirely exposed since the latest JLine 3.x or 4.x, the strict terminal closure check was introduced.
See jline3#1576 and jline3#1577

diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index f835061ac95002b92939e45a6f27ba0937d06c97..2283d3104f1425934e15ab38cee7ef823be5da83 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -483,6 +483,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
 
         this.hasFullyShutdown = true; // Paper - Improved watchdog support
+        io.papermc.paper.log.LoggerShutdown.shutdownLogging(); // Leaf - Improve logger/terminal shutdown
         System.exit(this.abnormalExit ? 70 : 0); // CraftBukkit // Paper - Improved watchdog support
     }
 
